src/orange/orangeinp/detail/LogicBuilderPolicies.hh

```diff
@@ -69,21 +72,69 @@ class PostfixLogicBuilderImpl
     inline void operator()(Joined const&);
     //!@}
 
+  protected:
+    VecLogic& logic() { return *logic_; }
+    ContainerVisitor<CsgTree const&, NodeId>& visite_node()
+    {
+        return visit_node_;
+    }
+
   private:
     ContainerVisitor<CsgTree const&, NodeId> visit_node_;
     VecSurface const* mapping_;
     VecLogic* logic_;
 };
 
+//---------------------------------------------------------------------------//
+/*!
+ * Recursively construct a logic vector from a node with infix operation.
+ *
+ * This is a policy used as template parameter of the \c
+ LogicBuilder::operator() . The
+ * user invokes this class with a node ID (usually representing a cell), and
+ * then this class recurses into the daughters using a tree visitor.
+ *
+ * Example: \verbatim
+    all(1, 3, 5) -> {{1, 3, 5}, "(0 & 1 & 2)"}
+    all(1, 3, any(~(2), ~(4))) -> {{1, 2, 3, 4}, "(0 & 2 & (~1 | ~3))"}
+ * \endverbatim
+ */
+class InfixLogicBuilderPolicy : public PostfixLogicBuilderPolicy
```

Comment:
I am not sure about inheritance vs independent implementation. We can avoid duplication by inheriting, but it can be error-prone as methods are not virtual, derived classes must make sure to reimplement base methods calling `operator()`, i.e. the NodeId, Negated, Joined, and Aliased overloads.

Reply:
You could have a base class `BuilderPolicyBase` that implements the shared methods and declares `operator()` and others to be pure virtual.

Reply:
I was trying to avoid using virtual functions, though this is only called on the host so this would be fine but CRTP can be a good alternative there.